cluster-autoscaler/FAQ.md

```diff
@@ -953,3 +957,190 @@ Caveats:
+vendor a same(or compatible) versions of `k8s.io`. Few times, you might have to revendor MCM into Autoscaler only due to the change of the `k8s.io` depedency in the autoscaler. 
+Please follow below steps to vendor new version of MCM:
+
+#### Step 1 (Optional)
+
+Vendor compatible `k8s.io` depedency into the MCM. 
+
+
+_Please make sure to vendor same `k8s.io` version into MCM, as the one already vendored in the autoscaler. As vendoring lower version of `k8s.io` into MCM, and later into the autoscaler may force entire autoscaler to use such low version of `k8s.io`. This can cause runtime issues if not build-issues._
+
+Update the `go.mod` file in the MCM to reflect expected versions of `k8s.io`. Then execute following:
+```
+cd <MCM_DIR>
+make revendor
+
+./hack/generate-code # regenerates the clients.
+```
+Autoscaler only vendors the `client` and `apis` related packages of the MCM, hence the `controller` package doesn't need to be adapted immediately in order to vendor MCM in Autoscaler.
+
+#### Step 2
+
+Update the [`go.mod-extra`](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/go.mod-extra) file to reflect the following:
+```
+require (
+  ...
+  github.com/gardener/machine-controller-manager vX.Y.Z # the new version
+)
+```
+
+If you are interested in vendoring MCM from the local-system or personal fork for testing, you can also add the `replace` section in `go.mod-extra` as shown below. 
+
+```
+// Replace the <$GOPATH> with the actual go-path. 
+replace (
+  ...
+  github.com/gardener/machine-controller-manager => <$GOPATH>/src/github.com/gardener/machine-controller-manager 
+
+  // OR you could also use the personal github-handle.
+  // github.com/gardener/machine-controller-manager => https://github.com/<USERNAME>/machine-controller-manager/tree/<BRANCH_NAME>
+)
+```
+
+#### Step 3:
+
+Setup the local-environment:
+```
+git clone https://github.com/gardener/autoscaler.git
+cd autoscaler/cluster-autoscaler
+git remote add origin git@github.com:gardener/autoscaler.git
+
+git checkout machine-controller-manager-provider
+git pull origin machine-controller-manager-provider
+
+git log | grep "Updating vendor against"  # Please save commit-hash from the first line of the output, it'll be used later, eg: 3eb90c19d0cf90b756c3e08e32c6495b91e0aeed
+
+``` 
+
+#### Step 4:
+
+This fork of the autoscaler vendors the [machine-controller-manager](https://github.com/gardener/machine-controller-manager) aka MCM from Gardener project. As the MCM itself vendors the `k8s.io` in it, we need to make following change to the [`update-vendor`](https://github.com/gardener/autoscaler/blob/master/cluster-autoscaler/hack/update-vendor.sh) script:
+
+Disable the check of implicit-depedencies of go.mod by commenting out following code in the update-vendor script.
+
+```
+#  if [[ "${IMPLICIT_FOUND}" == "true" ]]; then
+#    err_rerun "Implicit dependencies missing from go.mod-extra"
+#  fi
```

Comment:
Do we require this? I was able to vendor without this. Or maybe we can move it as optional step if you feel it's required at times.

Reply:
I made it `Optional` now.